File: frontend/src/lib/explorer/__tests__/statementClassGolden.test.ts
import { readFileSync } from "node:fs"
import { resolve } from "node:path"
import { describe, it, expect } from "vitest"
import { classifyExplorerStatement, type ExplorerStmtClass } from "../statementClass"

/**
 * Pins the UI classifier to the fixture the api-gateway's
 * validators.ClassifyStatementSQL is also pinned to
 * (statement_class_golden_test.go). The UI gates Run, the role message and the
 * destructive-confirm dialog on this class before the request is sent, so a
 * disagreement means the UI refuses statements the server would run, or offers
 * ones it will refuse.
 */
type GoldenCase = { sql: string; class: string }

const goldenPath = resolve(__dirname, "../../../../../shared/explorer_statement_class_golden.json")
const golden: GoldenCase[] = JSON.parse(readFileSync(goldenPath, "utf8"))

// The server calls a DML write "dml_write"; the UI calls it "write".
const toUiClass = (c: string): ExplorerStmtClass => (c === "dml_write" ? "write" : (c as ExplorerStmtClass))

describe("classifyExplorerStatement matches the server's shared golden", () => {
  it("golden fixture is not empty", () => {
    expect(golden.length).toBeGreaterThan(0)
  })

  it.each(golden.map((g) => [g.sql, toUiClass(g.class)] as const))("%j → %s", (sql, want) => {
    expect(classifyExplorerStatement(sql)).toBe(want)
  })
})

File: api-gateway/internal/validators/statement_class_golden_test.go
package validators

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// TestClassifyStatementSQLMatchesSharedGolden pins ClassifyStatementSQL to the
// fixture that frontend/src/lib/explorer/statementClass.ts is also pinned to
// (__tests__/statementClassGolden.test.ts). The UI copy gates the Run button,
// the role message and the destructive-confirm dialog before the request is
// sent, so when the two classifiers disagree the UI refuses statements the
// server would run, or offers ones it will refuse. A behaviour change here
// must update the fixture, which runs the TypeScript side too.
func TestClassifyStatementSQLMatchesSharedGolden(t *testing.T) {
	path := filepath.Join("..", "..", "..", "shared", "explorer_statement_class_golden.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read golden %s: %v", path, err)
	}
	var golden []struct {
		SQL   string `json:"sql"`
		Class string `json:"class"`
	}
	if err := json.Unmarshal(data, &golden); err != nil {
		t.Fatalf("parse golden: %v", err)
	}
	if len(golden) == 0 {
		t.Fatal("golden fixture is empty")
	}
	for _, g := range golden {
		if got := ClassifyStatementSQL(g.SQL); string(got) != g.Class {
			t.Errorf("ClassifyStatementSQL(%q) = %q, want %q", g.SQL, got, g.Class)
		}
	}
}
